File: scripts/enum/dynamodb.js
'use strict';

const {
  DynamoDBClient,
  ListTablesCommand,
  DescribeTableCommand,
  DescribeContinuousBackupsCommand,
  ListBackupsCommand,
  GetResourcePolicyCommand,
} = require('@aws-sdk/client-dynamodb');

const { withRetry, createLogger } = require('../lib');
const { baseEnum } = require('../lib/base-enum');

// --- Pagination for ListTables (uses ExclusiveStartTableName, not NextToken) ---

async function listAllTables(client, logger) {
  const allTables = [];
  let exclusiveStart = undefined;

  do {
    const params = {};
    if (exclusiveStart) {
      params.ExclusiveStartTableName = exclusiveStart;
    }

    logger.log('api_call', 'ListTables', { ExclusiveStartTableName: exclusiveStart || null });
    const response = await withRetry(() => client.send(new ListTablesCommand(params)));

    if (Array.isArray(response.TableNames)) {
      allTables.push(...response.TableNames);
    }

    exclusiveStart = response.LastEvaluatedTableName || null;
  } while (exclusiveStart);

  return allTables;
}

// --- Per-table enrichment ---

async function describeTable(client, tableName, logger) {
  logger.log('api_call', 'DescribeTable', { table: tableName });
  const response = await withRetry(() =>
    client.send(new DescribeTableCommand({ TableName: tableName }))
  );
  return response.Table;
}

async function describeContinuousBackups(client, tableName, logger) {
  logger.log('api_call', 'DescribeContinuousBackups', { table: tableName });
  try {
    const response = await withRetry(() =>
      client.send(new DescribeContinuousBackupsCommand({ TableName: tableName }))
    );
    const desc = response.ContinuousBackupsDescription;
    if (!desc) return null;
    return {
      continuous_backups_status: desc.ContinuousBackupsStatus || null,
      point_in_time_recovery:
        desc.PointInTimeRecoveryDescription?.PointInTimeRecoveryStatus === 'ENABLED',
    };
  } catch (err) {
    logger.log('warning', 'DescribeContinuousBackups_Failed', { table: tableName, error: err.message });
    return null;
  }
}

async function listBackups(client, tableName, logger) {
  logger.log('api_call', 'ListBackups', { table: tableName });
  try {
    const response = await withRetry(() =>
      client.send(new ListBackupsCommand({ TableName: tableName }))
    );
    return response.BackupSummaries ? response.BackupSummaries.length : 0;
  } catch (err) {
    logger.log('warning', 'ListBackups_Failed', { table: tableName, error: err.message });
    return 0;
  }
}

async function getResourcePolicy(client, tableArn, logger) {
  logger.log('api_call', 'GetResourcePolicy', { arn: tableArn });
  try {
    const response = await withRetry(() =>
      client.send(new GetResourcePolicyCommand({ ResourceArn: tableArn }))
    );
    if (response.Policy) {
      try {
        return JSON.parse(response.Policy);
      } catch {
        return response.Policy;
      }
    }
    return null;
  } catch (err) {
    // ResourceNotFoundException or AccessDeniedException — no policy exists or not supported
    if (err.name === 'ResourceNotFoundException' || err.name === 'PolicyNotFoundException' ||
        err.name === 'UnknownOperationException' || err.name === 'AccessDeniedException') {
      return null;
    }
    logger.log('warning', 'GetResourcePolicy_Failed', { arn: tableArn, error: err.message });
    return null;
  }
}

// --- Run (exported for testing) ---

async function run(opts = {}) {
  const runDir = opts.runDir;
  const region = opts.region;
  const accountId = opts.accountId;

  const logger = opts.logger || createLogger(runDir, 'dynamodb');
  let status = 'complete';
  const partialErrors = [];

  // DynamoDB client
  const client = opts.clients?.dynamodb ?? new DynamoDBClient({ region });

  // List all tables
  const tableNames = await listAllTables(client, logger);
  logger.log('info', 'TablesDiscovered', { count: tableNames.length });

  // Enumerate each table
  const findings = [];

  for (const tableName of tableNames) {
    try {
      const table = await describeTable(client, tableName, logger);

      // Encryption
      const sse = table.SSEDescription || {};
      const encryptionType = sse.SSEType || 'AES256'; // Default encryption if no SSE description
      const kmsKeyId = sse.KMSMasterKeyArn || null;

      // Streams
      const streamSpec = table.StreamSpecification || {};
      const streamEnabled = streamSpec.StreamEnabled || false;
      const streamViewType = streamSpec.StreamViewType || null;

      // Global table detection
      const isGlobalTable = !!(table.GlobalTableVersion || (table.Replicas && table.Replicas.length > 0));
      const replicas = (table.Replicas || []).map((r) => ({
        region: r.RegionName,
        status: r.ReplicaStatus,
      }));

      // Deletion protection
      const deletionProtection = table.DeletionProtectionEnabled || false;

      // Table class
      const tableClass = table.TableClassSummary?.TableClass || 'STANDARD';

      // Point-in-time recovery + continuous backups
      const backupInfo = await describeContinuousBackups(client, tableName, logger);

      // Manual backup count
      const backupCount = await listBackups(client, tableName, logger);

      // Resource policy
      const resourcePolicy = await getResourcePolicy(client, table.TableArn, logger);

      findings.push({
        resource_type: 'dynamodb_table',
        resource_id: tableName,
        arn: table.TableArn,
        region,
        encryption_type: encryptionType,
        kms_key_id: kmsKeyId,
        stream_enabled: streamEnabled,
        stream_view_type: streamViewType,
        is_global_table: isGlobalTable,
        replicas,
        deletion_protection: deletionProtection,
        table_class: tableClass,
        point_in_time_recovery: backupInfo ? backupInfo.point_in_time_recovery : null,
        continuous_backups_status: backupInfo ? backupInfo.continuous_backups_status : null,
        backup_count: backupCount,
        resource_policy: resourcePolicy,
        findings: [],
      });
    } catch (err) {
      partialErrors.push({ table: tableName, error: err.message });
      logger.log('error', 'TableEnumeration_Failed', { table: tableName, error: err.message });
    }
  }

  if (partialErrors.length > 0) {
    status = 'partial';
  }

  await logger.flush();
  return { findings, status };
}

if (require.main === module) {
  baseEnum({ module: 'dynamodb', run });
}

module.exports = { run };

File: enumerators/dynamodb.py
from __future__ import annotations

import json
from typing import Any

from scope_core.coverage import CoverageTracker
from scope_core.models import ModuleEnvelope


PRIMARY_CHECKS = ["list_tables"]
REQUIRED_CHECKS = ["describe_table"]
ABSENT_RESOURCE_POLICY_CODES = {
    "ResourceNotFoundException",
    "PolicyNotFoundException",
    "UnknownOperationException",
}


def run(factory: Any, region: str) -> ModuleEnvelope:
    tracker = CoverageTracker(primary=PRIMARY_CHECKS, required=REQUIRED_CHECKS)
    dynamodb = factory.client("dynamodb")
    findings: list[dict[str, Any]] = []

    try:
        table_names = _list_all_tables(dynamodb)
        tracker.record_ok("list_tables", resource="tables")
    except Exception as err:
        tracker.record_module_failure("list_tables", operation="dynamodb.ListTables", err=err)
        coverage, errors = tracker.to_envelope_fields()
        return ModuleEnvelope(
            module="dynamodb",
            account_id=factory.account_id,
            region=region,
            status=tracker.derive_status(),
            resources=[],
            coverage=coverage,
            errors=errors,
        )

    for table_name in table_names:
        try:
            table = dynamodb.describe_table(TableName=table_name).get("Table", {})
            tracker.record_ok("describe_table", resource=table_name)
        except Exception as err:
            tracker.record_resource_failure(
                "describe_table",
                resource=table_name,
                operation="dynamodb.DescribeTable",
                err=err,
            )
            continue

        table_arn = table.get("TableArn")
        finding = _base_table_finding(table, table_name, table_arn, region)

        try:
            response = dynamodb.describe_continuous_backups(TableName=table_name)
            description = response.get("ContinuousBackupsDescription") or {}
            pitr = description.get("PointInTimeRecoveryDescription") or {}
            finding["continuous_backups_status_value"] = description.get("ContinuousBackupsStatus")
            finding["point_in_time_recovery"] = pitr.get("PointInTimeRecoveryStatus") == "ENABLED"
            finding["continuous_backups_status"] = "present"
            tracker.record_ok("continuous_backups", resource=table_arn)
        except Exception as err:
            finding["continuous_backups_status"] = _classify_status(_error_code(err))
            tracker.record_skipped(
                "continuous_backups",
                resource=table_arn,
                reason=finding["continuous_backups_status"],
            )

        try:
            response = dynamodb.list_backups(TableName=table_name)
            finding["backup_count"] = len(response.get("BackupSummaries") or [])
            finding["backups_status"] = "present"
            tracker.record_ok("list_backups", resource=table_arn)
        except Exception as err:
            finding["backups_status"] = _classify_status(_error_code(err))
            tracker.record_skipped("list_backups", resource=table_arn, reason=finding["backups_status"])

        try:
            response = dynamodb.get_resource_policy(ResourceArn=table_arn)
            policy = response.get("Policy")
            if policy:
                finding["resource_policy"] = _parse_policy(policy)
                finding["resource_policy_status"] = "present"
            else:
                finding["resource_policy_status"] = "absent"
            tracker.record_ok("resource_policy", resource=table_arn)
        except Exception as err:
            code = _error_code(err)
            if code in ABSENT_RESOURCE_POLICY_CODES:
                finding["resource_policy_status"] = "absent"
                tracker.record_ok("resource_policy", resource=table_arn)
            else:
                finding["resource_policy_status"] = _classify_status(code)
                tracker.record_skipped("resource_policy", resource=table_arn, reason=finding["resource_policy_status"])

        findings.append(finding)

    coverage, errors = tracker.to_envelope_fields()
    return ModuleEnvelope(
        module="dynamodb",
        account_id=factory.account_id,
        region=region,
        status=tracker.derive_status(),
        resources=findings,
        coverage=coverage,
        errors=errors,
    )


def _list_all_tables(dynamodb: Any) -> list[str]:
    table_names: list[str] = []
    exclusive_start: str | None = None
    while True:
        kwargs = {"ExclusiveStartTableName": exclusive_start} if exclusive_start else {}
        response = dynamodb.list_tables(**kwargs)
        table_names.extend(response.get("TableNames") or [])
        exclusive_start = response.get("LastEvaluatedTableName")
        if not exclusive_start:
            return table_names


def _base_table_finding(table: dict[str, Any], table_name: str, table_arn: str | None, region: str) -> dict[str, Any]:
    sse = table.get("SSEDescription") or {}
    stream = table.get("StreamSpecification") or {}
    replicas = [
        {"region": replica.get("RegionName"), "status": replica.get("ReplicaStatus")}
        for replica in table.get("Replicas") or []
    ]
    table_class = (table.get("TableClassSummary") or {}).get("TableClass") or "STANDARD"

    return {
        "resource_type": "dynamodb_table",
        "resource_id": table_name,
        "arn": table_arn,
        "region": region,
        "encryption_type": sse.get("SSEType") or "AES256",
        "kms_key_id": sse.get("KMSMasterKeyArn"),
        "stream_enabled": stream.get("StreamEnabled") or False,
        "stream_view_type": stream.get("StreamViewType"),
        "is_global_table": bool(table.get("GlobalTableVersion") or replicas),
        "replicas": replicas,
        "deletion_protection": table.get("DeletionProtectionEnabled") or False,
        "table_class": table_class,
        "point_in_time_recovery": None,
        "continuous_backups_status_value": None,
        "continuous_backups_status": None,
        "backup_count": 0,
        "backups_status": None,
        "resource_policy": None,
        "resource_policy_status": None,
        "findings": [],
    }


def _parse_policy(policy: str) -> Any:
    try:
        return json.loads(policy)
    except json.JSONDecodeError:
        return policy


def _classify_status(code: str) -> str:
    return "access_denied" if code in {"AccessDenied", "AccessDeniedException"} else "error"


def _error_code(err: Exception) -> str:
    response = getattr(err, "response", None)
    if isinstance(response, dict):
        error = response.get("Error")
        if isinstance(error, dict) and error.get("Code"):
            return str(error["Code"])
    code = getattr(err, "Code", None)
    if code:
        return str(code)
    return err.__class__.__name__ or "Error"
